### DISCUSSION TITLE
TimerWheel.java why   long mask = SPANS[index] - 1;

### DISCUSSION THREAD

**Discussion by javakinglzh** [ben-manes/caffeine#541] [Q&A] [answered]
TimerWheel.java line 129 why   long mask = SPANS[index] - 1; why not   long mask = SPANS[index] - 1;
can not understand

**Comment by ben-manes** [comment, marked as answer]
Hi @javakinglzh,

You accidentally asked why `X` and not `X`? Can you please clarify.

In case it helps (but you might know this):
The mask is a cheap version of a modulus operand, which is otherwise a division instruction (a multi-cycle arithmetic operation). In the case of a power-of-two this can be replaced with a single cycle AND instruction, `x mod 2^i = x & (2^i – 1)`. This technique is popular in hash tables, including Java's, which improves performance due to how frequently they are used.

**Reply by javakinglzh** [reply]
sorry ,my mistake.
i want to know line 129 why long mask = SPANS[index] - 1; why not mask = timerWheel.length - 1;
so start = (int) (previousTicks & mask);
      end = 1 + (int) (currentTicks & mask);
and so line 134 and next & operattion is no need any more

**Reply by javakinglzh** [reply]
i can not understand why two times & the first is line 130  start = (int) (previousTicks & mask); and the second is line 136.
why just one time & : previousTicks & (timerWheel.length - 1)

**Reply by ben-manes** [reply]
We might need to step through with a debugger and unit tests. This code is a bit confusing due to using bitwise tricks for performance. There is `TimerWheelTest` but maybe more cases are needed.

A concern is that `System.nanoTime()` can overflow, so `currentTicks` might be less than `previousTimeNanos` in that scenario. That could cause a delay in removing some of the expired entries until the next full pass.

As ticks are the number of steps in the array, offhand I don't recall why a mask is needed at all.

In short, the code is too optimized that I can't be confident by sight alone and we would need to step through it. If there are improvements, you are welcome to send a PR.

**Reply by ben-manes** [reply]
I haven’t debugged yet, but I think it is because they would differ slightly. Assuming increasing values, the mask always has end larger than start. If using the array mask then end could be smaller as the wheels rotate. This would be fine, but needs the loop to handle the wrapping. In the given form, the loop increments and the array mask wraps for us. We’d get the same result with your mask if end was the absolute difference between the two values. We have that in the if condition.

I think the suggestion is equivalent and the same performance. I have no preference on which is used.

I probably ignored nano time overflowing as pragmatically won’t have much harm and wont happen in practice. That could be improved, though.

**Reply by javakinglzh** [reply]
> I haven’t debugged yet, but I think it is because they would differ slightly. Assuming increasing values, the mask always has end larger than start. If using the array mask then end could be smaller as the wheels rotate. This would be fine, but needs the loop to handle the wrapping. In the given form, the loop increments and the array mask wraps for us. We’d get the same result with your mask if end was the absolute difference between the two values. We have that in the if condition.
> 
> I think the suggestion is equivalent and the same performance. I have no preference on which is used.
> 
> I probably ignored nano time overflowing as pragmatically won’t have much harm and wont happen in practice. That could be improved, though.

I think it‘s a bug, let's compare the code and show an example .
If previousTimeNanos = 1152921503533105152 and currentTimeNanos is 1152921504606846976, previousTimeNanos is less than currentTimeNanos and  the span is 1073741824 （1.07s).
So the previousTick is  1073741823 and currentTick is 1073741824, previousTicks is small than currentTicks.
After code line 130 "**start = (int) (previousTicks & mask);**" the start is still 1073741823( 0x3FFFFFFF), however after code line 131 "**end = 1 + (int) (currentTicks & mask);**" , the end will be 1.
so the end will be less than start.
Am I clear? With my best honest.

**Reply by ben-manes** [reply]
Okay, would you like to write some more unit tests and try to fix the bugs?

**Reply by ben-manes** [reply]
From the discussion, something like this might be the fix then.

```java
void expire(int index, long previousTicks, long currentTicks) {
  Node<K, V>[] timerWheel = wheel[index];
  int mask = timerWheel.length - 1;

  int steps = 1 + (int) (currentTicks - previousTicks);
  int start = (steps < timerWheel.length)
      ? (int) (previousTicks & mask)
      : 0;

  for (int i = 0; i < steps; i++) {
    Node<K, V> sentinel = timerWheel[(start + i) & mask];
```

**Reply by javakinglzh** [reply]
Here is the test demo.

>    wheel.nanos = 1152921503533105152L; 
        wheel.schedule(new Timer(TimeUnit.NANOSECONDS.toNanos(1152921504606846976L)));
        System.out.println(wheel.toString());
        wheel.advance(TimeUnit.NANOSECONDS.toNanos(1152921517491748864L));
        System.out.println(wheel.toString());

Here is the result:

> Wheel #1: {0=[1152921504606846976]}
Wheel #2: {}
Wheel #3: {}
Wheel #4: {}
Wheel #5: {}
Wheel #1: {0=[1152921504606846976]}
Wheel #2: {}
Wheel #3: {}
Wheel #4: {}
Wheel #5: {}

It's clearly a bug. the advance function is useless

**Reply by ben-manes** [reply]
thanks. do you want to submit a PR or should I follow up on making a fix?

**Reply by ben-manes** [reply]
fyi, for your test case the proposed code changes passes.

**Reply by javakinglzh** [reply]
My computer has watchdog, I can not push code to github or submit PR.
You can make a fix.
With my best honest.

**Reply by ben-manes** [reply]
Thanks. I have a patch locally which handles this bug and for when the clock overflows. I'll try to wrap it up over the weekend or so, as its been hard to get focus for this project.

**Reply by ben-manes** [reply]
[timerWheel.patch.txt](https://github.com/ben-manes/caffeine/files/6386754/timerWheel.patch.txt)

**Reply by ben-manes** [reply]
Released in 2.9.1 and 3.0.2
